### PR TITLE
First release to use the I-NUCLEO-LRWAN1 expansion board as a LoRaWAN node in ABP or OTAA mode.

### PR DESCRIPTION
**Pull request by ghost**
Arduino library to use the I-NUCLEO-LRWAN1 expansion board as a LoRaWAN node in ABP or OTAA mode.

### CODE REVIEW

**Comment by LMESTM** [comment]
@fprwi6labs  starting review ...  can you please update the PR titles to something more relevant than 'add source file' ?

File: src/LoRaWANNode.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+  LoRaWANNodeClass();
+
+  bool begin(HardwareSerial *serialx, uint8_t band, uint8_t loraClass = LORA_CLASS_A);  //begin initialization function
+
+  bool joinOTAA(const char *appKey, const char *appEui = NULL); //function to configure the otaa parameters
+  bool joinABP(const char *devAddr, const char *nwkSKey, const char *appSKey); //function to configure the ABP parameters
+
+  int8_t sendFrame(char frame[], uint8_t length, bool confirmed, uint8_t port = 1);  //function to send a frame
+  bool receivedFrame(uint8_t *frame, uint8_t *length, uint8_t *port);
```
Comment (LMESTM) [inline src/LoRaWANNode.h]: why receivedFrame (in the past) rather than receiveFrame 
does it mean this function must be called after the Frame has been received?

File: src/LoRaWANNode.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+  // Lorawan stack & firmware version
+  void getVersion(String *str);
```
Comment (LMESTM) [inline src/LoRaWANNode.h]: can you briefly describe how to use this ? 
Who needs to allocate the string, caller of library or library itself ?

File: src/atcmd_modem.h [hunk 3]
```diff
@@ -152,85 +154,87 @@ typedef enum ATCmd
 #ifdef AT_CMD_STRING
 /*list of AT string cmd supported by the USI LoRa modem*/
 static char *CmdTab[] = {
-  {""},
-  {"I"},           /* firmware version of USI loara module*/
-  {"Z"},
-  {"+BAND"},       /* +BAND country band - default 868 band*/
+  "",
```
Comment (LMESTM) [inline src/atcmd_modem.h]: can you explain WHY you're removing all the { } ? 
this could maybe have been done in a specific commit.

File: src/debug.h [hunk 4]
```diff
@@ -58,62 +58,27 @@
 
 #include <string.h>
 #include <stdio.h>
-#include "hw_conf.h"
```
Comment (LMESTM) [inline src/debug.h]: Also explain why you need to clean-up debug.h and could also be done in a specific commit for easier tracking and file maintenance

File: src/hw_usart.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
+/******************************************************************************
+ * @file    hw_usart.c
```
Comment (LMESTM) [inline src/hw_usart.cpp:2]: Porting to Arduino Serial layer also could be done in specific commit to ease review and maintenance.

File: src/hw_usart.h [hunk 6]
```diff
@@ -47,39 +47,38 @@
 /* Define to prevent recursive inclusion -------------------------------------*/
 #ifndef __usart_H
 #define __usart_H
+
+// We define a larger UART RX buffer size to remove reception error (lost data).
```
Comment (LMESTM) [inline src/hw_usart.h]: Larger than what ?
where is SERIAL_RX_BUFFER_SIZE defined by default ?

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c [hunk 7]
```diff
@@ -173,97 +131,84 @@ ATEerror_t Modem_IO_Init( void )
...
   memset(LoRa_AT_Cmd_Buff, 0x00, sizeof LoRa_AT_Cmd_Buff);
 
   switch (at_group)
   {
-  case AT_CTRL:
-  {
-    Len = at_cmd_format( Cmd, NULL, CTRL_MARKER);
-    HAL_Status = at_cmd_send(Len);
-    if(HAL_Status != HAL_OK)
+    case AT_CTRL:
```
Comment (LMESTM) [inline src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: there seems to be only indentation or white spaces changes below - please it this kind of changes a separate commit to ease review and maintenance.

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c [hunk 8]
```diff
@@ -493,10 +474,18 @@ uint8_t value_8;   /*for 8_D*/
...
-HAL_StatusTypeDef RetCode;
+  HAL_StatusTypeDef RetCode;
+  uint16_t size;
+
+  DBG_PRINTF("cmd: %s\r\n", LoRa_AT_Cmd_Buff);
 
   /*transmit the command from master to slave*/
-  RetCode = HAL_UART_Transmit(&huart2, (uint8_t*)LoRa_AT_Cmd_Buff, len, 5000);
+  size = HW_UART_Modem_Write((uint8_t*)LoRa_AT_Cmd_Buff, len);
+  if(size == len ) {
```
Comment (LMESTM) [inline src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: if not all data has been sent, why not trying to send the remaining part ? 
something like
while (size < len ) {
size += HW_UART_Modem_Write((uint8_t*)LoRa_AT_Cmd_Buff + size, len - size);
}

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c [hunk 9]
```diff
@@ -509,93 +498,96 @@ HAL_StatusTypeDef RetCode;
...
+    if (/*(response[i] =='\r') || */(response[i] == '\n'))
+    {
+      DBG_PRINTF("at_cmd_receive: %s\r\n", response);
+      if(pdata == NULL) /*return code following a SET cmd or simple AT cmd*/
+      {
+        //if (i>1)  /*return code following a SET cmd or simple AT cmd- we skip the first <cr><ln>*/
+        //{
+          i= 0;
+          ResponseComplete = 1;
+          RetCode = at_cmd_responseAnalysing(&response[0]);
```
Comment (LMESTM) [inline src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: could be
at_cmd_responseAnalysing(response);
rather than 
at_cmd_responseAnalysing(&response[0]);

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c [hunk 10]
```diff
@@ -509,93 +498,96 @@ HAL_StatusTypeDef RetCode;
...
+        //}
+      }
+      else    /* returned value following a GET cmd */
+      {
+        if (i!= 0 && NoReturnCode)
         {
-          if (i>1)  /*return code following a SET cmd or simple AT cmd- we skip the first <cr><ln>*/
+          /*first statement to get back the return value*/
+          response[i] = '\0';
+          if (gFlagException != AT_FWVERSION)         /*see comment in loara_driver.c*/
```
Comment (LMESTM) [inline src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: is this only indentation changes again or have you made protocol / parsing changes against the default STM32 cube code ?
Mixing formatting changes with implementation adaptation will make the maintenance complex. I would prefer to avoid indentation and white spaces changes as much as possibe so that we can easily merge new version of the cube library if needed.

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.h [hunk 11]
```diff
@@ -52,13 +55,18 @@
 
 /* Exported types ------------------------------------------------------------*/
 
-#define DATA_RX_MAX_BUFF_SIZE    64       /*Max size of the received buffer*/
+#define DATA_RX_MAX_BUFF_SIZE    128       /*Max size of the received buffer*/
```
Comment (LMESTM) [inline src/i_nucleo_lrwan1_wm_sg_sm_xx.h]: why have you changes the DATA_RX_MAX_BUFF_SIZE ?

File: src/lora_driver.c [hunk 12]
```diff
@@ -44,16 +44,17 @@
  ******************************************************************************
  */
 
+#ifdef __cplusplus
+  extern "C" {
```
Comment (LMESTM) [inline src/lora_driver.c]: please explain why you are changing this lora_driver file for Arduino integration 
This should also be done in a separate commit and with only required modification.
Why removing all the code with #if USE_MDM32L07X01 ? 
this will make maintenance more complex

File: src/LoRaWANNode.cpp [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+/*
+ * @brief  Initializes the LoRa module
+ * @param  pointer to the serial instance uses to communicate with the module
+ * @param  LoRaWAN frequency band: 868MHz (EU) or 915MHz (US)
+ * @param  LoRaWAN node class: A, B or C. NOTE: Classes B and C are not available.
+ * @retval Error code: false if initialization failed else true
+ */
+bool LoRaWANNodeClass::begin(HardwareSerial *serialx, uint8_t band, uint8_t loraClass)
+{
+  uint8_t nbTry;
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp]: can be initialized to 0 here

File: src/LoRaWANNode.cpp [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+
+  if(!IS_BAND(band) || !IS_CLASS(loraClass) || (serialx == NULL)) {
+    return 0;
+  }
+
+  if(loraClass != LORA_CLASS_A) {
+    printf("The firmware only supports Class A so far.\r\n");
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp]: isn't there a DEBUG macro?

File: src/LoRaWANNode.cpp [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+  }
+
+  LoRa_DumyRequest();
+
+  //Echo mode must be disabled
+  Modem_AT_Cmd(AT_EXCEPT, AT_ATE, 0);
+
+  // Enable Lora module
+  nbTry = 0;
+  while(Lora_Init() != MODULE_READY) {
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp]: Why is there a need to make several trials ? is this something that happens frequently   ?

File: src/LoRaWANNode.cpp [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+ * @param  Network session key: 16 hexadecimal bytes
+ * @param  Application session key: 16 hexadecimal bytes
+ * @retval Error code: false if failed else true
+ */
+bool LoRaWANNodeClass::joinABP(const char *devAddr, const char *nwkSKey, const char *appSKey)
+{
+  uint8_t key[16];
+
+  if((devAddr != NULL) && (nwkSKey != NULL) && (appSKey != NULL)) {
+    keyCharToInt(devAddr, key, 4);
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp:133]: if keyCharToInt is only used in this file, could it be implemented locally ?

File: src/LoRaWANNode.cpp [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+/*
+ * @brief  Receives a LoRaWAN frame from the network. Must be called before the
+ *         next call of sendFrame(). The receive delay depends on the Rx1 and Rx2
+ *         windows.
+ * @param  pointer to the received frame.
+ * @param  Number of data received.
+ * @param  The application port number.
+ * @retval Error code: false if failed else true.
+ */
+bool LoRaWANNodeClass::receivedFrame(uint8_t *frame, uint8_t *length, uint8_t *port)
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp]: not clear, why receivedFrame() and not receiveFrame() ?

File: src/LoRaWANNode.cpp [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+  if(str != NULL) {
+    char tmp[10] = {'\0'};
+    Lora_GetVersion((uint8_t *)tmp);
+    str->concat(tmp);
+  }
+}
+
+/* NOTE: verbose mode is enabled when the 'ATI' command is sent. But it is
+  impossible to remove it (even if +VERB=0 is sent??!!). The verbose mode must be
+  disable if not the AT_VSSCANF function doesn't work.
```
Comment (LMESTM) [inline src/LoRaWANNode.cpp]: Comment unclear. please clarify. Does it mean that Verbose mode cannot be used ? Why ?

File: examples/LoRaPingPong/LoRaPingPong.ino [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    radio.write(PingMsg, 4);
+    timer = millis();
+  }
+
+  if(radio.read(rcvData) > 0) {
+    if(memcmp(rcvData, PongMsg, 4) == 0) {
+      Serial.println((char *)PongMsg);
+    } else if(memcmp(rcvData, PingMsg, 4) == 0) {
+      isMaster = false;
+      radio.write(PongMsg, 4);
```
Comment (VVESTM) [inline examples/LoRaPingPong/LoRaPingPong.ino]: This is working well. For code comprehension, I would have prefer to print what we send and what we receive. For example : 
  if(radio.read(rcvData) > 0) {
    Serial.print("rcvData = ");Serial.println((char *)rcvData);
    if(memcmp(rcvData, PingMsg, 4) == 0) {
      isMaster = false;
      Serial.print("send = ");Serial.println((char *)PongMsg);
      radio.write(PongMsg, 4);
    }
  }

File: examples/LoRaPingPong/LoRaPingPong.ino [hunk 20]
```diff
@@ -0,0 +1,62 @@
+/*
+  LoRaPingPong.ino
+
+  Provides an example to use the shield as a simple LoRa radio.
+  Data are exchange between two boards. The first receiving the PING message
+  becomes slave and will just send a PONG answer.
+
```
Comment (VVESTM) [inline examples/LoRaPingPong/LoRaPingPong.ino:7]: This is not limited to 2 boards. All boards in the area can see the ping pong messages.

File: src/LoRaWANNode.h [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+  /* enter/exit the sleep mode */
+  bool sleep(void);
+  bool wakeup(void);
+
+private:
+
+};
+
+extern LoRaWANNodeClass node;
```
Comment (LMESTM) [inline src/LoRaWANNode.h]: maybe LoraNode rather than node ? (or lora ...or loranode ...)

File: examples/LoRaWANABP/LoRaWANABP.ino [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+// Data send
+char frameTx[] = "Hello world!";
+
+void setup()
+{
+  Serial.begin(9600);
+  Serial.println("-- LoRaWAN ABP sketch --");
+
+  // Enable the USI module and set the radio band.
+  while(!node.begin(&SerialLora, LORA_BAND_EU_868)) {
```
Comment (LMESTM) [inline examples/LoRaWANABP/LoRaWANABP.ino]: as commented in previous commit, maybe lora rather than node would make it more clear?

File: examples/LoRaWANABP/LoRaWANABP.ino [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+  int status = node.sendFrame(frameTx, sizeof(frameTx), UNCONFIRMED);
+  if(status == LORA_SEND_ERROR) {
+    Serial.println("Send frame failed!!!");
+  } else if(status == LORA_SEND_DELAYED) {
+    Serial.println("Module busy or duty cycle (not sent)");
+  } else {
+    Serial.println("Frame sent");
+
+    // Check if data received from a gateway
+    if(node.receivedFrame(frameRx, &len, &port)) {
```
Comment (LMESTM) [inline examples/LoRaWANABP/LoRaWANABP.ino]: isn't there a specified timeout to this call ?

File: README.md [hunk 24]
```diff
@@ -1,2 +1,47 @@
...
+
+Arduino library to support I-NUCLEO-LRWAN1 LoRa® expansion board based on USI®
+LoRaWAN™ technology module.
+
+## API
+
+This library provides an Arduino API to manage the I-NUCLEO-LRWAN1 expansion
+board as a LoRaWAN™ node.
+
+* You be able to choose the ABP or OTAA mode and set the corresponding keys.
```
Comment (LMESTM) [inline README.md]: You be able -> You will be able

File: src/LoRaRadio.cpp [hunk 25]
```diff
@@ -0,0 +1,456 @@
+/**
```
Comment (LMESTM) [inline src/LoRaRadio.cpp:1]: Please provide more info in commit message
what is LORA RAdio API meant for ?

File: src/LoRaRadio.cpp [hunk 26]
```diff
@@ -0,0 +1,456 @@
...
+        return (AT_TEST_PARAM_OVERFLOW);
+      }
+    }
+    i++;
+  }
+
+  return ( RetCode);   /*version of HAL .. there was not Rx field state*/
+}
+
+LoraRadio radio;
```
Comment (LMESTM) [inline src/LoRaRadio.cpp]: radio ->loraradio ?

**Review by LMESTM — COMMENTED** [review]
Overall looks good to me.
Few questions and comments raised for clarity and for easier maintenance

File: examples/LoRaPingPong/LoRaPingPong.ino [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+*/
+
+#include "LoRaRadio.h"
+
+#define SEND_PERIOD_MS 1000 //send period every second.
+
+// Serial port use to communicate with the USI shield. Specific to the board Discovery L475 IoT.
```
Comment (VVESTM) [inline examples/LoRaPingPong/LoRaPingPong.ino:14]: Comment can be more explicit. This shield is difficult to use with Nucleo boards as the debug USART use the same pins... If user needs to use this on Nucleo, some soldering is needed on the Nucleo (see UM1724, 6.8 section)
A comment can also be added in the Readme file.
 
By the way, I also tested with a Disco F746 and this is not working. Don't know why yet.

**Review by VVESTM — COMMENTED** [review]
Hi,
Some comments on tests.
I was able to execute PingPong and OTAA tests. No ABP mode for now, need to check how to configure gateway.

File: library.properties [hunk 28]
```diff
@@ -0,0 +1,9 @@
+name=I-NUCLEO-LRWAN1
```
Comment (VVESTM) [inline library.properties]: Add stm32duino in the name to be aligned with other libraries

Comment (fpistm) [reply library.properties]: name=STM32duino I-NUCLEO-LRWAN1

File: examples/LoRaWANABP/LoRaWANABP.ino (hunk 23 above)
Comment (ghost) [reply examples/LoRaWANABP/LoRaWANABP.ino]: The timeout is done inside this method in the AT command reception mechanism.

File: src/LoRaWANNode.cpp (hunk 15 above)
Comment (ghost) [reply src/LoRaWANNode.cpp]: Because after disable the echo mode the first AT answer doesn't match with the format expected.

File: src/LoRaWANNode.cpp (hunk 16 above)
Comment (ghost) [reply src/LoRaWANNode.cpp:133]: Not only used in this file.

File: src/LoRaWANNode.cpp (hunk 18 above)
Comment (ghost) [reply src/LoRaWANNode.cpp]: Verbose mode can't be used because the AT command driver provides by ST expects a specific format because the AT_VSSCANF function is a tiny implementation and is not able to parse some characters. Verbose mode must be off.

File: src/LoRaWANNode.h (hunk 1 above)
Comment (ghost) [reply src/LoRaWANNode.h]: Renamed.

File: src/atcmd_modem.h (hunk 3 above)
Comment (ghost) [reply src/atcmd_modem.h]: To remove compilation warnings.
It's already done in a specific commit.

File: src/debug.h (hunk 4 above)
Comment (ghost) [reply src/debug.h]: It's already done in a specific commit.

File: src/hw_usart.h (hunk 6 above)
Comment (ghost) [reply src/hw_usart.h]: Defined in HardwareSerial.h. SERIAL_RX_BUFFER_SIZE is set to 64 by default but it is too small for our use case.

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c (hunk 7 above)
Comment (ghost) [reply src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: White spaces are already clean in a previous commit.

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.c (hunk 8 above)
Comment (ghost) [reply src/i_nucleo_lrwan1_wm_sg_sm_xx.c]: This code comes from by the ST library. I kept the logic used.

File: src/i_nucleo_lrwan1_wm_sg_sm_xx.h (hunk 11 above)
Comment (ghost) [reply src/i_nucleo_lrwan1_wm_sg_sm_xx.h]: I faced to a memory leak issue.

File: src/lora_driver.c (hunk 12 above)
Comment (ghost) [reply src/lora_driver.c]: Modification done in a specific commit.
I removed all USE_MDM32L07X01 to be coherent. Only one target per library as the other libraries provided.

File: examples/LoRaPingPong/LoRaPingPong.ino (hunk 20 above)
Comment (ghost) [reply examples/LoRaPingPong/LoRaPingPong.ino:7]: Of course but I'm not sure how this sketch will react with more than 2 boards.

File: examples/LoRaPingPong/LoRaPingPong.ino (hunk 27 above)
Comment (ghost) [reply examples/LoRaPingPong/LoRaPingPong.ino:14]: Comment added in "known limitations" section of the readme.

File: examples/LoRaPingPong/LoRaPingPong.ino (hunk 19 above)
Comment (ghost) [reply examples/LoRaPingPong/LoRaPingPong.ino]: You will not be able to know which is master or slave because you will print all messages received from the network.

File: examples/LoRaWANABP/LoRaWANABP.ino (hunk 23 above)
Comment (LMESTM) [reply examples/LoRaWANABP/LoRaWANABP.ino]: don't you think it may be useful to have a timeout that can be defined by the application ? 
Or is the timeout related to the LORA protocol  ? If the timeout needs to be kept inside receivedFrame, then it is needed to well describe the timeout value in the API.

File: src/LoRaWANNode.cpp (hunk 16 above)
Comment (LMESTM) [reply src/LoRaWANNode.cpp:133]: Ok ... I haven't seen other locations where is used ...

File: src/LoRaWANNode.cpp (hunk 15 above)
Comment (LMESTM) [reply src/LoRaWANNode.cpp]: This cannot be guessed so please explain this in detail here.

File: src/LoRaWANNode.cpp (hunk 18 above)
Comment (LMESTM) [reply src/LoRaWANNode.cpp]: OK. Please complete the existing comment with more details

File: examples/LoRaWANABP/LoRaWANABP.ino (hunk 23 above)
Comment (ghost) [reply examples/LoRaWANABP/LoRaWANABP.ino]: By default the driver didn't implement a timeout inside at_cmd_receive_async_event_downlink_data(). It was a blocking function so I add a timeout to exit after 5 seconds. This allows to be sure to receive an answer inside the RX1 or RX2 windows.

File: src/LoRaWANNode.cpp (hunk 15 above)
Comment (ghost) [reply src/LoRaWANNode.cpp]: Sometimes if the module is not ready when we call `Modem_AT_Cmd(AT_EXCEPT, AT_ATE, 0);` the first answer of the command `Modem_AT_Cmd(AT_CTRL, AT, NULL );` inside `LoraInit()` is not "OK" but "AT OK" (echo of the command sent). The driver doesn't understand the answer in this case. So I used a loop to avoid this issue.

File: src/LoRaWANNode.h (hunk 2 above)
Comment (ghost) [reply src/LoRaWANNode.h]: String must be allocate at sketch level (caller of library).

File: src/LoRaWANNode.cpp (hunk 15 above)
Comment (LMESTM) [reply src/LoRaWANNode.cpp]: When I say "here" I mean in the comments of the file

File: src/LoRaWANNode.h (hunk 2 above)
Comment (LMESTM) [reply src/LoRaWANNode.h]: I finally understood it, but that must be explained along the API

File: src/LoRaRadio.cpp (hunk 25 above)
Comment (LMESTM) [reply src/LoRaRadio.cpp:1]: My comment above has not been addressed

**Comment by LMESTM** [comment]
@fprwi6labs  I  haven't done a full review, but I still find comments of mine that have not been covered.
I asked to view some explanations in the commit messages.  Typically what is Lora radio API and why it is being added.  See https://chris.beams.io/posts/git-commit/

I'd like to see commit message BODY to be able to understand the changes by reading the commits.

**Comment by ghost** [comment]
@LMESTM 
commit message body added for LoRa API.

@VVESTM 
ABP and OTAA examples modified. Some messages in reception were lost when messages sent were postponed because of the duty cycle.

File: src/debug.h (hunk 4 above)
Comment (LMESTM) [reply src/debug.h]: I don't agree - there are several different modifications, with several rationals behind, all into this single commit:
Adapt STM32CubeExpansion_LRWAN_V1.1.2 driver files to Arduino
there is warning clean-up in strings, debug modifications ( I still don't know why debug was modified), all the hw_ layer seems to be modified, DATA_RX_MAX_BUFF_SIZE  has been modified  without a good explanation etc.
Each modification has a different objective and should be described in a commit with a good description.  This is needed for a more efficient maintenance of the files.

Comment (ghost) [reply src/debug.h]: Commits reworked as requested.

**Comment by ghost** [comment]
Add RX1 delay configuration method in the LoRaWAN API.
Add explanation in OTAA and ABP sketches about the keys configuration.
Add advice in readme about the default configuration of the module.

**Comment by ghost** [comment]
* Add a method to the API to configure the Join Request RX1 delay.
* Fix OTAA join request retransmission failure. I think the source of the problem is inside the USI module. I leave you try and confirm this patch is satisfactory.

**Review by fpistm — APPROVED** [review]
All examples successfully tested with a Multitech Conduit AEP gateway and Discovery L475 IoT with I-NUCLEO-LRWAN1 LoRa® expansion board. Several gateway configuration tested (Public/private network with different RxJoin delay,...)
Only keyword.txt is missing
